# Optimisation de portefeuille d'actions



# Membres du groupe:
---



> Guelfane Abdelaziz    (Implémentation et simulations numériques)

> Reda Benkirane    (Analyse mathématique)

> Ismail Oulkhir    ( description de la problématique et du phénomène modélisé & formulation et Analyse mathématique)

> Omar Bellmir    (Analyse mathématique)

> Imad Absri    (Formulation mathématique sous forme d'un problème d'optimisation bien posé)




## 1 Présentation de la problématique<a id="0"></a>

Le projet d'Optimisation de Portefeuille est une initiative qui vise à maximiser les rendements financiers tout en minimisant les risques associés à la gestion d'un portefeuille d'investissement. Dans le monde des marchés financiers, les investisseurs sont constamment à la recherche de stratégies permettant de générer des profits tout en protégeant leur capital contre les fluctuations et les incertitudes du marché. Ainsi on se confronte aux problèmes suivant :
> Comment modéliser les prix des actifs ainsi que les portefeuilles ?

>Comment peut-on exprimer le problème de recherche d'un portefeuille performant en respectant un seuil maximal de risque spécifié ?

>Comment traduire mathématiquement et résoudre le problème en utilisant
des outils d'optimisation ?

## 2 Formulation Mathématique<a id="1"></a>

On suppose le marché à 2 dates t = 0 et t = 1.

Il existe N actifs risqués, et 1 actif sans risque.

L'actif sans risque vaut 1 en t = 0 et 1+r en t = 1.

On note $p_i(t)$ le prix de l'actif i à la date t, avec $i ∈ \{1, ..., N \}$ et $t ∈ \{0, 1\}$
On note $y_i$ le rapport $\frac{p_i(1)}{p_i(0)}$.
- Afin de modéliser les valeurs futures des actifs, on utilise des variables aléatoires, pour cela on se donne un espace de probabilité $(\Omega, \mathcal{F}, \mathbb{P})$
- $R=\left(\begin{array}{c}r_1 \\ \vdots \\ r_N\end{array}\right)$ vecteur aléatoire à valeurs dans $\mathbb{R}^N$ qu'on suppose dans $L^2$. Ce vecteur représente le rendement des investiments dans les actifs donnés.
- On note son espérance
$$
\mu=\mathbb{E}[R] \in \mathbb{R}^N
$$
et sa matrice de variance covariance
$$
\Omega=\operatorname{Var}(Y)=\left(\mathbb{C o v}\left(Y_i, Y_j\right)_{1 \leq i, j \leq N}\right) \in \mathbb{S}_n^{++}(\mathbb{R})
$$
Donc en particulier $\Omega$ inversible.
- On suppose que $\forall i \in\{1, \ldots, N\}$ on a $\mu_i>1+r$
- Un portefeuille est définit par le couple $\left(x_0, x\right) \in \mathbb{R} \times \mathbb{R}^N$, indiquant la quantité d'actif qu'il contient.
- La valeur du portefeuille aux différents instants est donc
$$
\left\{\begin{array}{l}
V_0\left(x_0, x\right)=x_0+\sum_{i=1}^N x_i p_i(0) \\
V_1\left(x_0, a\right)=x_0(1+r)+\sum_{i=1}^N x_i p_i(1)=x_0(1+r)+\sum_{i=1}^N x_i p_i(0) r_i
\end{array}\right.
$$
Sous forme matricielle:
Valeurs du portefeuille
$$
\left\{\begin{array}{l}
V_0\left(x_0, x\right)=x_0+x^T p(0) \\
V_1\left(x_0, x\right)=x_0(1+r)+x^T \operatorname{diag}(p(0)) R
\end{array}\right.
$$
où
$$
p(0)=\left(\begin{array}{c}
p_1(0) \\
\vdots \\
p_N(0)
\end{array}\right)
$$



### Formulation mathématique sous forme d'un problème d'optimisation bien posé:
 - On considère un portefeuille de valeur $v$ en $t=0$, i.e $V_0\left(a_0, a\right)=v$, on se donne un niveau maximal de variance $\sigma^2$, on veut naturellement maximiser l'espérance de gain.
- On cherche à trouver les couples $\left(a_0, a\right)$ qui résolvent:
$$
\begin{array}{ll}
\max _{a_0, a} & \mathbb{E}\left[V_1\left(a_0, a\right)\right] \\
\text { sous } & \operatorname{Var}\left[V_1\left(a_0, a\right)\right] \leq \sigma^2 \\
& V_0\left(a_0, a\right)=v
\end{array}
$$

## 3 Analyse Mathématique du problème d'optimisation<a id="1"></a>

On note
$$
w_x= diag(p(0))x = \left(\begin{array}{c}
p_1(0)*x_1 \\
\vdots \\
p_N(0)* x_N
\end{array}\right)
$$

Donc
$$
\left\{\begin{array}{l}
V_0\left(x_0, x\right)=x_0+ w_x^T e \\
V_1\left(x_0, a\right)=x_0(1+r)+w_x^TR
\end{array}\right.
$$

où $e = (1 … 1)^T$
En général pour un vecteur aléatoire $X ∈ L^2$ et un vecteur réel $λ$ de même
dimension, $Var(λT X) = λ^T Var(X)λ$.

On en déduit :
$$
\left\{\begin{array}{l}
\mathbb{E}(V_1(x_0,x))=x_0(1+r)+w_x^T\mu \\
Var(V_1(x_0,x))=w_x^T\Omega w_x
\end{array}\right.
$$

Une forme équivalente de ce problème d'optimisation est :
  \begin{array}{ll}
\max _{w_x \in \mathbb{R}^N} & w_x^T\tilde{μ} \\
\text { sous } & w_x^T\Omega w_x \leq \sigma^2 \\
\end{array}

où $\tilde{μ} = \mu -(1+r)e $




#Condition de Salter
Pour un problème convexe différentiable les conditions de KKT sont
nécessaires et suffisantes pour l'optimalité, en d'autre terme un point x est
optimal pour le primal si on réussit à trouver λ et μ de sorte que (x, λ, μ)
satisfasse les conditions de KKT.

###Mise sous forme de problème de minimisation sous contrainte


La forme canonique d'un problème d'optimisation sous contraintes sur $x \in \mathbb{R}^N$ est :
  \begin{array}{ll}
\min &J(x) \\
\text { sous contraintes } & f_i(x) \leq 0,&  i=1\dots,m \\
&h_i(x)=0,&j=1,\dots,p
\end{array}
Avec toutes les fonctions  à valeurs dans $\mathbb{R}.$

En réécrivant le problème équivalent sous problème de minimisation:
  \begin{array}{ll}
-\min _{w_x \in \mathbb{R}^N} & -w_x^T\tilde{μ} \\
\text { sous } & w_x^T\Omega w_x \leq \sigma^2 \\
\end{array}
En appliquant ce qu'on a vu en cours pour l'optimisation sous contraintes:

$J(x)= -w_x^T\tilde{μ} $ est linéaire donc convexe.

$f_1(x)=w_x^T\Omega w_x -\sigma^2$ qui est convexe car $\nabla^2 f_1(x)=2\Omega \in \mathbb{S}_n^{++}(\mathbb{R}) $

Dans le cas $\sigma=0$ , on investit tout dans l'actif sans risque.


#Existence et Unicité
Dans le cas  $\sigma \ge 0$ , en prenant x=0 on trouve que $w_x^T\Omega w_x   =0 \le \sigma^2 $ et donc 0 satisfait la condition de Slater : $f_{1}(0)=0 \le 0$.
Il existe donc une et une seule solution optimale pour le problème d'optimisation convexe différentiable, sa forme analytique sera donnée par la suite.

###Conditions d'optimalités
 et donc les conditions de KKT s'écrivent:

>$w_x^T\Omega w_x -\sigma^2 \leq 0$ (1)

>$ \lambda \geq 0$ (2)

>$\lambda(w_x^T\Omega w_x -\sigma^2)=0 $ (3)

>$-\tilde{\mu}+\lambda\Omega x=0$ (4)

sont nécessaires et suffisantes pour l'optimalité de ce problème convexe.

La condition 3 implique $ \lambda=0 $ ou $w_x^T\Omega w_x -\sigma^2=0$

Si $\lambda =0 $ alors la condition 4 donne $\tilde{\mu} =0 $ ou $\mu>(1+r)e$ donc $\tilde{\mu} \ge 0$ d'ou $\lambda \ge 0$ et c'est l'autre terme qui s'annule.

Cela fait sens, la variance du portefeuille de gain maximal atteindra la
variance maximale permise $\sigma^2$.

L'équation 4 donne $x=\frac{1}{\lambda}\Omega^{-1}\tilde{\mu}$

En réinjectant tout cela dans les équations initiales on trouve l'expression pour notre solution :
        \begin{array}{l}
x^*\left(\lambda^*\right)=\frac{1}{\lambda^*} \operatorname{diag}(p(0))^{-1} \Omega^{-1} \tilde{\mu} \\
x_0^*\left(\lambda^*\right)=v-\frac{1}{\lambda^*} \tilde{\mu}^T \Omega^{-1} e \\
\lambda^*=\frac{1}{\sigma} \sqrt{\tilde{\mu}^T \Omega^{-1} \tilde{\mu}}
\end{array}

## Sélection d'un algorithme de résolution



 # <center> Méthode du gradient projeté

L'objectif d'optimisation de portefueille consiste à maximiser les rendements financiers. La méthode du gradient projeté permet de rechercher efficacement cette maximisation en propjetant la solution du problème d'optimisation sur l'espace des solutions admissivbles défini par $K = \{x \in \mathbb{{R}}^n,  x^T\Omega x \leq \sigma^2\}$, garantissant ainsi que la solution respecte les contraintes imposées.


Nous décrivons maintenant la méthode du gradient projeté. Comme son nom l’indique, on cherche à résoudre un problème sous contraintes
$$ \arg\min\{J(x), x \in K\} $$


où $K = \{x \in \mathbb{{R}}^n,  x^T\Omega x \leq \sigma^2\}$


par une méthode de type descente de gradient. Le problème est que le gradient de $J$ peut nous faire
sortir de $K$. L’idée est donc de considérer le projeté du gradient. Les itérations sont de la forme (ici pour un gradient projeté à pas constant)

Nonobstant, la méthode du gradient projeté s’inspire de la méthode du
gradient.

Dans le cas sans contraintes, l’algorithme du gradient, qui est une méthode de descente, s’écrit sous la forme générique :
$$
\left\{\begin{array}{l}
\ x_0  \in \mathbb{{R}}^n \\
x_{k+1}=x_k + ρ_kd_k, d_k\in \mathbb{{R}}^n-\{0_{\mathbb{{R}}^n}\}, ρ_k \in \mathbb{R^{+*}}
\end{array}\right.
$$

$ρ_k$ et $d_k$ sont choisis de telle sorte que $J(x_k+ρ_kd_k)\leq J(x_k)
$
, cependant, c'est là où réside le problème, c'est lorsqu’on minimise sur  l'ensemble de contraintes $K$ et que $x_k ∈ K$ on n’est pas sûr avec la formulation précédente que l’itéré $x_{k+1}=x_k + ρ_kd_k$ appartienne à $K$.

Il sera donc nécessaire de  "ramener" l'élément dans $K$, ce qui est réalisé en effectuant une projection sur $K$.

C'est  grâce à cette propriété qu'on a pu distinguer les choix parmi les autres algorithmes, et opter pour la méthode du gradient projeté.

On réalise cette dernière opération grâce à une projection sur $K$. On appelle projection sur $K$ l'application $P_K \colon \mathbb{{R}}^n \to \mathbb{R}$ définie par
<center> $$P_K(x):=\arg\min\{\left\|k-x\right\|, k \in K\} $$
















##Algorithme de la méthode du gradient projeté

**1. Initialisation:** $k=0:$ choix de $x_0$ et de $\rho_0>0$

**2. Iteration $k$  :**   $x_{k+1}=\ P_K\left(x_k-\rho_k \nabla J\left(x_k\right)\right)$;

**3. Critère d'arrêt:** Si $\left\|x_{k+1}-x_k\right\|<\varepsilon$, Break



### Convergence de l'algorithme

On a la fonction objective $C^1$ de $\mathbb{R}^n$ dans $\mathbb{R}$, le minimum existe et est unique, puisque dans ce cas la fonction est linéaire, le gradient projeté converge vers le point minimum.

## 4 Implémentation et simulations numériques <a id="1"></a>

In [ ]:
import numpy as np
from scipy.optimize import minimize

def Gradient_projete(mu_tilde, omega, sigma_carre, x0, alpha, tol=1e-6, max_iter=1000):
    x = x0
    traj = [x]

    def projection_func(x):
        return np.linalg.norm(x - x0)

    for iter in range(max_iter):
        x_old = x

        # Prédiction / descente de gradient
        grad = -mu_tilde
        x = x - alpha * grad

        # Projection sur l'ensemble K
        constraint = {'type': 'ineq', 'fun': lambda x: x.T.dot(omega).dot(x) - sigma_carre}
        bounds = [(None, None)] * len(x)
        result = minimize(projection_func, x, method='SLSQP', constraints=constraint, bounds=bounds)
        x = result.x

        err = np.linalg.norm(x_old - x)

        traj.append(x)

        # Vérification de la convergence
        if err < tol:
            print('L\'algorithme a convergé après', iter + 1, 'itérations.')
            break


    return x


In [ ]:
# Exemple d'utilisation
mu_tilde = np.array([0.1, 0.2, 0.15])  # Rendements espérés des actifs
omega = np.array([[0.04, 0.02, 0.01], [0.02, 0.09, 0.03], [0.01, 0.03, 0.05]])  # Matrice de variance-covariance
sigma_carre = 0.04  # Variance maximale autorisée
x0 = np.array([0.2, 0.3, 0.5])  # Répartition initiale des actifs dans le portefeuille
alpha = 0.1  # Paramètre de pas
epsilon = 0.001  # Critère d'arrêt

optimal_valeurs = Gradient_projete(mu_tilde, omega, sigma_carre, x0, alpha, tol=1e-6, max_iter=1000)
print("Répartition optimale des actifs dans le portefeuille :")
for i, valeur in enumerate(optimal_valeurs):
    print(f"Actif {i+1}: {valeur}")

L'algorithme a convergé après 3 itérations.
Répartition optimale des actifs dans le portefeuille :
Actif 1: 0.21072255886537458
Actif 2: 0.3261626320727165
Actif 3: 0.5201645683943722


In [ ]:
import cvxpy as cp
import numpy as np

def optimize_portfolio(mu, omega, sigma_squared, tolerance=1e-6):
    n = len(mu)  # Nombre d'actifs

    # Variables de décision
    x = cp.Variable(n, nonneg=True)  # Répartition des actifs dans le portefeuille

    # Problème d'optimisation
    objective = cp.Maximize(mu @ x)
    constraints = [x @ omega @ x <= sigma_squared, cp.sum(x) == 1]

    # Résolution du problème d'optimisation avec compteur
    problem = cp.Problem(objective, constraints)
    iterations = 0  # Compteur d'itérations
    prev_solution = None  # Solution précédente
    while True:
        problem.solve()
        if problem.status != cp.OPTIMAL:
            break
        iterations += 1
        if prev_solution is not None and np.linalg.norm(x.value - prev_solution) <= tolerance:
            break
        prev_solution = x.value.copy()

    # Récupération de la solution optimale
    optimal_valeurs = x.value

    return optimal_valeurs, iterations

# Exemple d'utilisation
mu = np.array([0.1, 0.2, 0.15])  # Rendements espérés des actifs
omega = np.array([[0.04, 0.02, 0.01], [0.02, 0.09, 0.03], [0.01, 0.03, 0.05]])  # Matrice de variance-covariance
sigma_carre = 0.04  # Variance maximale autorisée

optimal_valeurs, iterations = optimize_portfolio(mu, omega, sigma_carre)
print("Répartition optimale des actifs dans le portefeuille :")
for i, valeur in enumerate(optimal_valeurs):
    print(f"Actif {i+1}: {valeur}")
print("Nombre d'itérations nécessaires :", iterations)

Répartition optimale des actifs dans le portefeuille :
Actif 1: 0.15434123083664117
Actif 2: 0.40521482323524116
Actif 3: 0.4404439459280821
Nombre d'itérations nécessaires : 2


### Comparaison des resultats des deux algorithmes : précision et vitesse de convergence /nombre d'iterations

In [ ]:
import cvxpy as cp
import numpy as np
from scipy.optimize import minimize
import time
import matplotlib.pyplot as plt


# Exemple d'utilisation et comparaison des méthodes
mu_tilde = np.array([0.1, 0.2, 0.15])  # Rendements espérés des actifs
omega = np.array([[0.04, 0.02, 0.01], [0.02, 0.09, 0.03], [0.01, 0.03, 0.05]])  # Matrice de variance-covariance
sigma_carre = 0.04  # Variance maximale autorisée

mu_tilde = -mu
sigma_carre = sigma_carre
x0 = np.array([0.2, 0.3, 0.5])  # Répartition initiale des actifs dans le portefeuille
alpha = 0.1  # Paramètre de pas

# Méthode 1 : Gradient Projeté
start_time_1 = time.time()
optimal_valeurs_1 = Gradient_projete(mu_tilde, omega, sigma_carre, x0, alpha, tol=1e-6, max_iter=1000)
execution_time_1 = time.time() - start_time_1

# Méthode 2 : cvxpy
start_time_2 = time.time()
tolerance=1e-6
optimal_valeurs_2 = optimize_portfolio(mu, omega, sigma_carre)
execution_time_2 = time.time() - start_time_2

# Comparaison des résultats
print("Méthode Gradient Projeté - Répartition optimale des actifs dans le portefeuille :")
for i, valeur in enumerate(optimal_valeurs_1):
    print(f"Actif {i+1}: {valeur}")

print("Méthode cvxpy - Répartition optimale des actifs dans le portefeuille :")
print("Répartition optimale des actifs dans le portefeuille :")
for i, valeur in enumerate(optimal_valeurs):
    print(f"Actif {i+1}: {valeur}")



# Affichage des temps d'exécution
print("Temps d'exécution - Méthode Gradient Projeté:", execution_time_1, "secondes")
print("Temps d'exécution - Méthode cvxpy:", execution_time_2, "secondes")


Méthode Gradient Projeté - Répartition optimale des actifs dans le portefeuille :
Actif 1: 0.21072277420563215
Actif 2: 0.32616551551503425
Actif 3: 0.5201607118310778
Méthode cvxpy - Répartition optimale des actifs dans le portefeuille :
Répartition optimale des actifs dans le portefeuille :
Actif 1: 0.15434123083664117
Actif 2: 0.40521482323524116
Actif 3: 0.4404439459280821
Temps d'exécution - Méthode Gradient Projeté: 2.7804183959960938 secondes
Temps d'exécution - Méthode cvxpy: 0.013319015502929688 secondes


On observe des différences entre les deux méthodes en termes de répartition des actifs. On voit clairement que la résultats relatifs aux actifs sont légèrement différents, le temps d'exécution pour la méthode du **gradient projeté** est de 2.78 secondes, tandis que le temps de convergence pour la méthode **cvxpy** est de 0.013 seconde. Donc en termes de vitesse de convergence, la méthode cvxpy parait être plus rapide que celle du gradient projeté.

### Discussion et interpretation des résultats trouvés

Au vu des résultats trouvés au préalable, la méthode du gradient projeté et de cvxpy peuvent utiliser différentes formulations pour résoudre le problème d'optimisation. Cela inclut les contraintes, les objectifs et les algorithmes spécifiques utilisés. Les différences dans la formulation peuvent conduire à des solutions légèrement différentes. De plus, les 2 méthodes sont relativement différentes. Le gradient projeté est basé sur la descente de gradient avec une projection sur l'ensemble des contraintes ($K$ dans notre cas), tandis que cvxpy peut utiliser des méthodes d'optimisation convexe plus sophistiquées et spécialisées. Ces différences en elles-mêmes peuvent créer des résultats pratiquement différents.

Les légères différences peuvent bien entendu être aussi attribuées à des erreurs numériques ou à des approximations numériques dans les calculs effectués par les 2 méthodes, notamment, pour la méthode du gradient projeté, car on choisit manuellement la tolérance (seuil d'arrêt), le pas de descente (alpha) ou autre. Cela étant dit, ces calculs numériques peuvent introduire de petites variations dans les résultats obtenus.

Il est aussi important de souligner que la méthode utilisée par cvxpy est plus rapide dans notre cas, cette méthode est conçu pour exploiter les propriétés du problème d'optimisation spécifique et trouver rapidement la solution.
En outre, la méthode du gradient projeté est basée sur une itération de la projection du gradient, ce qui peut nécessiter un plus grand nombre d'itérations pour converger vers la solution optimale. Tandis que la méthode de cvxpy est basée sur une optimisation convexe qui permet d’assurer une convergence plus rapide et cela ne peut être justifiée que par les résultats trouvés. Nonobstant, il se peut, dans cetains cas, que la méthode du gradient projeté soit plus optimale en terme de précision de la solution par rapport à l'autre méthode, malgré la rapidité de cette dérnière.

## 5 Conclusion <a id="1"></a>

L’ optimisation du portefeuille grâce à la théorie moderne du portefeuille peut être un moyen efficace d’obtenir des rendements plus élevés et une meilleure gestion des risques.

Toutefois, il est important de noter que la théorie moderne des portefeuilles présente certains inconvénients, notamment le fait que les portefeuilles sont évalués en fonction de la variance plutôt que du risque de baisse. Cela signifie que même si l’optimisation du portefeuille peut améliorer le rendement global, elle ne protège pas nécessairement contre les événements extrêmes du marché ou les risques de baisse.

Bien que la théorie moderne des portefeuilles comporte des limites, elle demeure un outil très utile pour les investisseurs qui cherchent à diversifier leurs portefeuilles et à améliorer la gestion des risques.



> "*Tous les modèles sont erronés, mais certains sont utiles.*" ~ George Box

> "*Le talent remplit les pages, mais la persévérance complète les chapitres.*" ~ C.J. Cherryh


